### PR TITLE
Avoid pulling unnecessary data with Cassandra getRange 

### PR DESCRIPTION
**Pull request by nziebart**
**Goals (and why)**:
Fixes https://github.com/palantir/atlasdb/issues/2479

When performing a range scan, we can simply grab the latest version for each column for each row, rather than pulling all the data for each row.

**Implementation Description (bullets)**:
- If only 1 column is requested in a range scan, fetch just the latest version of that column for each row
- Some refactoring of the slice predicate creation logic
- Tweak some of the existing tests for better coverage here
- Add a TODO to optimize the multi-column case

**Concerns (what feedback would you like?)**:
- Are there any edge cases here?

**Where should we start reviewing?**:
`CassandraKeyValueServiceImpl#getRangeWithPageCreator `

**Priority (whenever / two weeks / yesterday)**:
This is a big perf improvement, so let's get it in soon, but also want to make sure this is correct.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2480)

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java [hunk 1]
```diff
@@ -1476,7 +1483,22 @@ CfDef getCfForTable(TableReference tableRef, byte[] rawMetadata, int gcGraceSeco
...
             ConsistencyLevel consistency,
             Supplier<ResultsExtractor<T>> resultsExtractor) {
-        RowGetter rowGetter = new RowGetter(clientPool, queryRunner, consistency, tableRef, ColumnFetchMode.FETCH_ALL);
+        SlicePredicate predicate = new SlicePredicate();
+        if (rangeRequest.getColumnNames().size() == 1) {
+            byte[] col = rangeRequest.getColumnNames().iterator().next();
+            ByteBuffer start = CassandraKeyValueServices.makeCompositeBuffer(col, timestamp - 1);
+            ByteBuffer end = CassandraKeyValueServices.makeCompositeBuffer(col, -1);
+            SliceRange slice = new SliceRange(start, end, false, 1);
+            predicate.setSlice_range(slice);
```
Comment (nziebart) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java]: @clockfort i'm struggling to grasp why this was never implemented this way before. there don't seem to be any edge cases here because we literally discard every value except the most recent anyway... but yet i feel like there must be some reason we'd make such an obviously inefficient query here

**Comment by fsamuel-bs** [comment]
Note that we're relying on the fact that the timestamps in C* are written in decreasing order with this change. Even though I think this was always true, it would still be good to double check.

Also worth checking what happens when we put and then putUnlessExists and then put again. We may write 0 in the ts column, which would always make us fetch the putUnlessExists value? I guess we filter the value with the `ts < startTs` of the `ValueExtractor#internalExtractResult`, but I may be wrong here.

@jboreiko for SA.

**Comment by tpetracca** [comment]
> we're relying on the fact that the timestamps in C* are written in ordered from the bigger 

This is 100% true

> put and then putUnlessExists and then put again

This is unsupported and bad practice. We removed (or deprecated?) the `TABLE.putUnlessExsits` because it was already incorrect.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java (hunk 1 above)
Comment (tpetracca) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java]: @nziebart maybe it was a limitation in Cassandra 1.2.X? CLocks would know more concretely though.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java [hunk 2]
```diff
@@ -1493,19 +1480,28 @@ CfDef getCfForTable(TableReference tableRef, byte[] rawMetadata, int gcGraceSeco
...
-            long timestamp,
+            long startTs,
             ConsistencyLevel consistency,
             Supplier<ResultsExtractor<T>> resultsExtractor) {
+        SlicePredicate predicate;
+        if (rangeRequest.getColumnNames().size() == 1) {
+            byte[] colName = rangeRequest.getColumnNames().iterator().next();
+            predicate = SlicePredicates.latestVersionForColumn(colName, startTs);
+        } else {
+            // TODO(nziebart): optimize fetching multiple columns by performing a parallel range request for
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java]: Is this work going to happen in this PR, a separate PR, or to be tracked as an issue?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java [hunk 3]
```diff
@@ -1493,19 +1480,28 @@ CfDef getCfForTable(TableReference tableRef, byte[] rawMetadata, int gcGraceSeco
...
+            // each column. note that if no columns are specified, it's a special case that means all columns
+            predicate = SlicePredicates.create(Range.ALL, Limit.NO_LIMIT);
+        }
         RowRangeLoader rowRangeLoader = new RowRangeLoader(clientPool, queryRunner, consistency, tableRef);
         ColumnGetter columnGetter = new ThriftColumnGetter();
 
-        return getRangeWithPageCreator(
-                rowRangeLoader, ColumnFetchMode.fetchAll(), columnGetter, rangeRequest, resultsExtractor, timestamp);
+        return getRangeWithPageCreator(rowRangeLoader, predicate, columnGetter, rangeRequest, resultsExtractor,
+                startTs);
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java]: Let's also change the `timestamp` parameter of the below method to `startTs`.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/paging/ColumnFetchMode.java [hunk 4]
```diff
@@ -1,36 +0,0 @@
...
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-package com.palantir.atlasdb.keyvalue.cassandra.paging;
-
-public final class ColumnFetchMode {
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/paging/ColumnFetchMode.java]: 🎉

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java [hunk 5]
```diff
@@ -454,13 +455,7 @@ private void lowerConsistencyWhenSafe() {
                             @Override
                             public Map<Cell, Value> apply(Client client) throws Exception {
                                 // We want to get all the columns in the row so set start and end to empty.
```
Comment (fsamuel-bs) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java]: I think this comment is not needed anymore.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/thrift/SlicePredicates.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+        Range ALL = Range.of(UNBOUND_START, UNBOUND_END);
+
+        static Range singleColumn(byte[] columnName, long maxTimestampExclusive) {
+            ByteBuffer start = startOfColumn(columnName, maxTimestampExclusive);
+            ByteBuffer end = endOfColumn(columnName);
+            return of(start, end);
+        }
+
+        static ByteBuffer startOfColumn(byte[] columnName, long maxTimestampExclusive) {
```
Comment (fsamuel-bs) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/thrift/SlicePredicates.java]: I would rename this to `startColumn`. Every `(columnName, ts)` pair is a Cassandra column.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java (hunk 2 above)
Comment (nziebart) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java]: created https://github.com/palantir/atlasdb/issues/2498

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/thrift/SlicePredicates.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import java.nio.ByteBuffer;
+
+import org.apache.cassandra.thrift.SlicePredicate;
+import org.apache.cassandra.thrift.SliceRange;
+import org.immutables.value.Value;
+
+import com.palantir.atlasdb.encoding.PtBytes;
+import com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServices;
+
+public final class SlicePredicates {
```
Comment (fsamuel-bs) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/thrift/SlicePredicates.java]: We should document that a SlicePredicate just limits the column selection of the query, as the documentation for the thrift api is lacking in this. We could do this by renaming all our `SlicePredicate` variables to `columnPredicates`, or encapsulating a SlicePredicate on a `ColumnSlicePredicate`. If we don't want to do either of those, we could add Javadocs to this class explaining this.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/paging/ColumnFetchMode.java (hunk 4 above)
Comment (nziebart) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/paging/ColumnFetchMode.java]: yeah this class was pretty weird. mainly the name

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/thrift/SlicePredicates.java (hunk 6 above)
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/thrift/SlicePredicates.java]: I don't think I would. It's the start of the particular column - `startColumn` would imply to me that `columnName` was the first of many columns.

**Comment by nziebart** [comment]
So we do have a benchmark that will get this - `KvsGetRangeDirty`

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/thrift/SlicePredicates.java (hunk 6 above)
Comment (fsamuel-bs) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/thrift/SlicePredicates.java]: But this is exactly the case with `getRowsColumnRangeForSingleHost`. See CassandraKVSImpl#794

**Comment by gsheasby** [comment]
+1 pending test run of the benchmark...

**Comment by nziebart** [comment]
Ran some benchmarks on my timelock benchmarks branch (yet unmerged). These were performed after a `nodetool flush` + `nodetool compact`

(range scan of 1000 rows, each row = 1KB)
develop:
0 overwrites: 32ms
10 overwrites: 207ms
100 overwrites: 2169ms

this branch:
0 overwrites: 30ms
10 overwrites: 46ms
100 overwrites: 160ms

So looks like the benefits are fairly linear, as we'd expect. Scans still slow down a bit with overwrites, likely since we have to do more disk seeking, but the impact is much smaller.

**Comment by nziebart** [comment]
Also notably, if I run with 16 parallel threads, we start seeing timeouts and failures (after 6 attempts) for the 100 overwrite test on develop.

This branch succeeds with a mild p50 increase from 160ms to 300ms

**Comment by gsheasby** [comment]
Sounds good - let's add a release note documenting the measured perf improvement, and we can merge this.
